packages/toc/src/toc_item.tsx

```diff
@@ -57,9 +119,77 @@ class TOCItem extends React.Component<IProperties, IState> {
       event.stopPropagation();
       heading.onClick();
     };
+    /**
+     * Callback invoked upon clicking a menu item to run code cells.
+     */
+    const runCells = () => {
+      let code: INotebookHeading[];
+      let h: INotebookHeading;
+      let i: number;
+
+      // Find the heading in the list of ToC headings...
+      const headings = toc;
+      for (i = 0; i < headings.length; i++) {
+        if (heading === headings[i]) {
+          break;
+        }
+      }
+      code = [];
+
+      // Check if the current heading is a "code" heading...
+      h = heading as INotebookHeading;
+      if (h.type === 'code') {
+        code.push(h);
+      }
+      // Find all nested code headings...
+      else {
+        const level = heading.level;
+        for (i = i + 1; i < headings.length; i++) {
+          h = headings[i] as INotebookHeading;
+          if (h.level <= level) {
+            break;
+          }
+          if (h.type === 'code') {
+            code.push(h);
+          }
+        }
+      }
+      // Run each of the associated code cells...
+      for (i = 0; i < code.length; i++) {
+        if (code[i].cellRef) {
+          const cell = code[i].cellRef as CodeCell;
+          const panel = cell.parent?.parent as NotebookPanel;
+          if (panel) {
+            void CodeCell.execute(cell, panel.sessionContext);
+          }
+        }
+      }
+    };
+
+    const onContextMenu = (event: React.MouseEvent) => {
+      event.preventDefault();
+      const commands = new CommandRegistry();
+      const contextMenu = new Menu({ commands });
+      commands.addCommand(CommandIDs.runCells, {
+        execute: runCells,
+        label: 'Run Cells'
+      });
+      contextMenu.addItem({ command: CommandIDs.runCells });
```

Comment:
It might be possible to define the context menu only once rather than at each event.

Reply:
Agreed :)

Reply:
+1 and I did some refactoring :)